Refreshing Tags Editor: Mark 2

jleandroperez opened the pull request:
### Description:
In this PR we're refreshing the way Tags are rendered in the Tags Editor field.

@aerych I'm starting to get the hang out of AppKit. I THINK!!

Mind taking a peek at this one sir? (Thank you!!!)

Ref. #458 

### Details:
There is no (public) way to customize **NSTokenField's** UX. Internally, this component is similar to NSTextView, and relies on a (private) NSTextAttachment subclass to draw the bubbles.

In order to customize the way NSTokenField bubbles are rendered, we're:

1. Implementing our own *NSTokenFieldCell* subclass: **TagsFieldCell**
2. Injecting, in few key spots, a mechanism to swap the (private) NSTextAttachments cell with our own implementation

**Note:** Cool thing? in AppKit, whenever you enter Edition mode, the SDK replaces the component with a NSTextView instance (called the Field Editor). 

In this PR we're hooking up to that mechanism, too, to customize the way Tokens are rendered during edition (everything is contained in **TagsFieldCell**.

---

### Test: UX
- [x] Verify that Tags in the TagsEditor area are **vertically centered**, and properly spaced
- [x] Verify that the Tag Token's Baseline (**text position.y**) matches the TagsEditor's Baseline
- [x] Verify the Tags Editor reacts well to **Dark Mode <> Light Mode** transitions

### Test: Editing
- [x] Verify that new tags are added as soon as you type a new keyword + press **return**
- [x] Verify you can select tags and delete them with **delete**
- [ ] Verify that pressing **ESC** while editing a tag causes the TagsField to resign first responder
- [ ] Verify that Editing a second tag (while you were editing another one) causes the former to get tokenized (bubble wrapped!)
- [ ] Verify that Selecting a second tag (while you were editing another one) causes the former to get tokenized (bubble wrapped!)

### Test: Mouse
- [x] Verify that **clicking** over any Tag causes its BG to be highlighted
- [x] Verify that **double clicking** over any Tag allows you to edit the Tag's Text
- [x] Verify that words can be dragged from the **TextView** into the **TagsEditor**
- [ ] Verify that Tags cannot be dragged from the **TagsEditor** into the **TextView**

### Release
`RELEASE-NOTES.txt` was updated in 0fab90d with:
 
> Refreshed the Tags Editor Interface!

### Screenshots
<img width="500" src="https://user-images.githubusercontent.com/1195260/84688032-fe5dfa80-af14-11ea-8710-76a10b418145.jpg">

mokagio commented:
@jleandroperez this PR has milestone 1.11, which we aim to code freeze this upcoming Monday, 2020/06/15, but is still in draft.

Do you want it in, or should it slide to the next release?

jleandroperez commented on Simplenote/BackgroundView.swift:
```diff
@@ -22,6 +22,15 @@ class BackgroundView: NSView {
+    @IBInspectable
+    var drawsTopBorder: Bool = false {
+        didSet {
+            needsDisplay = true
```
Thank you!!! totally missed this!

jleandroperez commented on Simplenote/TagAttachmentCell.swift:
```diff
@@ -0,0 +1,135 @@
+    override func cellSize() -> NSSize {
+        let textSize    = nsStringValue.size(withAttributes: attributes)
+        let width       = textSize.width.rounded(.up) + Metrics.textInsets.left + Metrics.textInsets.right + Metrics.bgInsets.left + Metrics.bgInsets.right
+        let height      = textSize.height.rounded(.up) + Metrics.textInsets.top + Metrics.textInsets.bottom + Metrics.bgInsets.top
```
You're veeeery right, missed this!! thank you for pointing this out!

jleandroperez commented on Simplenote/TagAttachmentCell.swift:
```diff
@@ -0,0 +1,135 @@
+        drawText(in: cellFrame)
+    }
+
+    override func draw(withFrame cellFrame: NSRect, in controlView: NSView?, characterIndex charIndex: Int, layoutManager: NSLayoutManager) {
```
**Now that you mention this!!** I ended up leaving the first implementation that I managed to get working.

Both API(s) are since 10.10 ([Here](https://developer.apple.com/documentation/appkit/nstextattachmentcellprotocol/1508412-draw) and [here](https://developer.apple.com/documentation/appkit/nstextattachmentcellprotocol/1508385-draw).

As far as I understand, the SDK simply calls the "simplest override available" (extract from NSTextAttachment.h):

```
The class NSTextAttachmentCell implements them to simply call the simpler methods
```

In the aims of... keeping things as simple as possible, I'm switching to the API that doesn't receive a LayoutManager reference (since we don't need it). And adding notes to detail exactly when `draw` method runs (Edition or Display modes).

**Thank you Eric!!**

jleandroperez commented:
PR Test Steps updated (up there!).

1. Added `ESC` support (it'll "commit" whatever you were doing)
2. Multiple Tags Edition / Selection is now disallowed
3. Reworked the TagAttachment mouse support (moved the handling to TagsField!)

@aerych ready for a second look!! **Thank you!!!**

aerych commented:
Looks great @jleandroperez !  :shipit:

jleandroperez commented:
Woooo thank you Eric!!!
